Feature: Read Later

alfaarghya opened the pull request:
## This PR contains the Read Later feature (#28) for **Bulletty**

### **Updates:**
1. config `$DATA_DIR/.later.tml` to store the Read Later entries
2. _Icon_: The articles that are added to the `Read Later` section will have an icon beside their title.
3. _Read Later Section_: 
    - IF no files are added to the `Read Later` section, it will be hidden until the user adds an article to `Read Later`.
    - `Read Later` has a counter `()` that will keep track of how many articles are added.
4. _Key Binds_:
    - **Add Article to Read Later**:  On `MainScreen`, press `L` to add.
    - **Remove Article from Read Later**:  On `ReaderScreen`, press `L` to remove.
    - **Help**: On each screen, press `?` to check available `Read Later` keybinds
5. Increase the size of `instructions_text` in `statusline`. Causenewly added instructions are getting off the screen.

Note - I use Copilot for suggestions and for some explanations.

### Preview
https://github.com/user-attachments/assets/1d0e32fe-55ef-4d2d-b296-6d3eeafe397c

---

I hope this PR will enhance the functionality of **bulletty.**

CrociDB commented on src/core/library/feedlibrary.rs:
```diff
@@ -178,11 +179,44 @@ impl FeedLibrary {
+        }
+    }
+
+    pub fn is_in_read_later(&self, url: &str) -> bool {
```
Why check for URL and not the path? If you keep the list of paths in `.later.toml`, you don't even need to load any entry in order to do most of the operations like this.

CrociDB commented on src/ui/screens/readlaterscreen.rs:
Oh no @alfaarghya, don't worry about that! No problem at all. :)

I just say a lot of "don't waste my time" because I see a lot of people are just trying to profit by submitting half-assed vibe-coded PRs, I can see you're really working on it and learning on the process!

alfaarghya commented:
> ## .later.toml storage
> 
> I see you're storing all the info for the entry in the `.later.toml` file. I think it's better to store only the list of paths to the entries rather than duplicated information. I know that it initially it means way more file oprations, since you'd have to open all those files to gather their information, but I think it's a way more concise storage system. These files can be synced by the user and must be really simple to also read, or interpreted by other programs that would like to implement the same set of functionalities as bulletty.
> 
> In the future, I want to implement some sort of local cache (that is not in the DATA folder) that will make a big different in terms of file system access and make it perform a lot better without changing the user data.


Sure thing, I will update it!

alfaarghya commented:
> ## List separator
> 
> My idea for the item separator in the list is that it wouldn't be selectable. But I honestly think it's a minor issue now, so don't need to worry changing it, I can take care of it later. Unless you really want to do it. 😋
> <img alt="image" width="205" height="65" src="https://private-user-images.githubusercontent.com/70588/502020821-9460692f-7ebb-4526-ae17-0532c0caa3f3.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjA2MjMzMzUsIm5iZiI6MTc2MDYyMzAzNSwicGF0aCI6Ii83MDU4OC81MDIwMjA4MjEtOTQ2MDY5MmYtN2ViYi00NTI2LWFlMTctMDUzMmMwY2FhM2YzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTEwMTYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUxMDE2VDEzNTcxNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTNjZWFlZjFjYjU4YjRkMDIzZTVmNTdlMGViMjcyNDU2NzE1MWZhNTlhNjZhMWQzYzBlZjMwNjQxYmU4YmM0ZjImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.PP_F7GGzr1SEKwIdTVCAm69jtXK4yrElPVBBji549Ws">

Should I remove it then?

CrociDB commented:
Keep the List Separator the way it is, @alfaarghya! I'll later just make it non-selectable, but for now that is really good!

alfaarghya commented:
> ### Removal not happening properly
> 
> I noticed that after adding and removing a lot of entries to the Read Later section, the `.later.toml` file was looking like this:
> <img alt="image" width="885" height="757" src="https://private-user-images.githubusercontent.com/70588/502496455-ce38035c-c820-4997-9824-1c25a4404c90.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjA2OTM0ODcsIm5iZiI6MTc2MDY5MzE4NywicGF0aCI6Ii83MDU4OC81MDI0OTY0NTUtY2UzODAzNWMtYzgyMC00OTk3LTk4MjQtMWMyNWE0NDA0YzkwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTEwMTclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUxMDE3VDA5MjYyN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWU2NWVkYWU4MDYzYzg4NGUzY2NlODgyOWQwZTQ3NjY4ZTgwZjQwM2M1NDZiZTI2YTFjYjk5NWVlMjA3MDlmYzEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.Ymn2JGTs6dtO4ocaCQIi2O1jp1Io2AlbSQCnJkfkLzo">
> 
> A lot of "empty" entries.

---

#### I just checked it, and this behaviour is not happening on my system.
<img width="1920" height="1080" alt="Screenshot From 2025-10-17 14-59-59" src="https://github.com/user-attachments/assets/52ee8d75-3221-4d86-8b3e-53cd9a93e7d6" />

### Can you guide me on how I can reproduce this behaviour?

CrociDB commented on src/core/library/data/librarydata.rs:
```diff
@@ -331,6 +333,120 @@ impl LibraryData {
+        }
+
+        let contents = std::fs::read_to_string(&read_later_path)?;
+        let data: ReadLaterData = toml::from_str(&contents)
```
Maybe it's a good idea to cache this list, for all the `is_in_read_later` calls. Only update it when we need to load or save.

CrociDB replied:
By cache I just mean a local cache, like having a field `pub read_later_list: Vec<String>` in the `LibraryData` struct

CrociDB commented on src/core/library/data/librarydata.rs:
```diff
@@ -331,6 +333,120 @@ impl LibraryData {
+    }
+
+    pub fn get_read_later_feed_entries(&self) -> color_eyre::Result<Vec<FeedEntry>> {
+        let data = self.load_read_later()?;
```
May be a little bit of a picky thing, but I saw you always call the read later list `data`, which is very generic and could lead to confusion. Maybe you could call it `read_later_list` or something like that.

CrociDB commented on src/core/library/data/librarydata.rs:
```diff
@@ -331,6 +333,120 @@ impl LibraryData {
+
+    fn absolute_path_to_relative_path(&self, file_path: &str) -> String {
+        let rel_path = if let Some(categories_pos) = file_path.find("categories/") {
+            let prefix_to_strip = &file_path[..categories_pos + "categories/".len()];
```
Also, I think there's a way to get the relative path of an absolute by subtracting from another absolute path, which I think is more safe. Something like getting `file_path - self.path.join(DATA_CATEGORIES_DIR)`.

alfaarghya replied:
Ok I will check it

CrociDB commented:
Hey @alfaarghya, interesting. I deleted the `.later.toml` file and tried again and it didn't do anymore. Maybe it was a leftover from how the system was before? No problem then, scratch that! :)

alfaarghya commented on src/core/library/readlaterentry.rs:
```diff
@@ -0,0 +1,54 @@
+use crate::core::feed::feedentry::FeedEntry;
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct ReadLaterEntry {
```
Successfully removed it :)

alfaarghya commented:
### Hii @CrociDB, I implemented all of the requested changes and here are some basic details -

### Updates:

1. Remove the `ReadLaterEntry` completely and keep only the file path
2. Implemented the `read_later_cache` to reduce the file operation on `load_read_later()`. Due to cache implementation, I need to update the mutability of some function parameters.
3. Use `DATA_CATEGORIES_DIR` in `absolute_path_to_relative_path()` and update over all logic.
4.  update variable name => `data` -> `read_later_list`

### Fix:

Remove Dead file entries from `.later.toml`

alfaarghya commented:
Thanks @CrociDB : )

I could not have done it without your guidance. Thanks for it!
